stdlib/LinearAlgebra/src/blas.jl

```diff
@@ -107,29 +107,65 @@ end
 openblas_get_config() = strip(unsafe_string(ccall((@blasfunc(openblas_get_config), libblas), Ptr{UInt8}, () )))
 
 """
-    set_num_threads(n)
+    set_num_threads(n::Integer)
+    set_num_threads(::Nothing)
+
+Set the number of threads the BLAS library should use equal to `n::Integer`.
+If `nothing` is passed to this function, julia tries to figure out the optimial number of threads.
+The exact heuristic is an implementation detail.
 
-Set the number of threads the BLAS library should use.
+On exotic variants of `BLAS` this function can fail, which is indicated by returning `nothing`.
 """
 function set_num_threads(n::Integer)
     blas = vendor()
-    if blas === :openblas
-        return ccall((:openblas_set_num_threads, libblas), Cvoid, (Int32,), n)
-    elseif blas === :openblas64
-        return ccall((:openblas_set_num_threads64_, libblas), Cvoid, (Int32,), n)
+    if blas === :openblas || blas == :openblas64
+        return ccall((@blasfunc(openblas_set_num_threads), libblas), Cvoid, (Int32,), n)
     elseif blas === :mkl
         # MKL may let us set the number of threads in several ways
         return ccall((:MKL_Set_Num_Threads, libblas), Cvoid, (Cint,), n)
-    end
-
-    # OSX BLAS looks at an environment variable
-    @static if Sys.isapple()
+    elseif Sys.isapple()
+        # OSX BLAS looks at an environment variable
         ENV["VECLIB_MAXIMUM_THREADS"] = n
+    else
+        @warn "Failed to set number of BLAS threads." maxlog=1
     end
 
     return nothing
 end
 
+function set_num_threads(::Nothing)
+    n = max(1, Sys.CPU_THREADS ÷ 2)
+    set_num_threads(n)
+end
```

Comment:
I'm a bit uneasy about `Sys.CPU_THREADS ÷ 2` (e.g., the user may have hyperthreading disabled). But I'm not sure how much this code path would be used...

Also, it makes sense to use whatever the logic used in `Base.__init__`

https://github.com/JuliaLang/julia/blob/35c1f87176fecab5a8e8077fb3a62275363a5aa5/base/Base.jl#L380-L388

So, maybe it's better to read the environment variable (e.g., if the user or  `Base.__init__` sets this environment variable, it's going to be used in `set_num_threads`).

```suggestion
_tryparse_env_cint(key) = tryparse(Cint, get(ENV, key, ""))

function set_num_threads(::Nothing)
    n = something(
        _tryparse_env_cint("OPENBLAS_NUM_THREADS"),
        _tryparse_env_cint("OMP_NUM_THREADS"),
        max(1, Sys.CPU_THREADS ÷ 2),
    )
    set_num_threads(n)
end
```

(or maybe without `÷ 2`)

Reply:
I think the fallback to ` max(1, Sys.CPU_THREADS ÷ 2)` is good as a last resort until julia gets functionality for determining number of physical cores (Who knows this might be sooner than later)